### PR TITLE
libcamera: init at 1.0.0

### PR DESCRIPTION
**Pull request by RealityAnomaly**
###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 129387` at 0772a02b run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built successfully:</summary>
  <ul>
    <li>libcamera</li>
  </ul>
</details>
<details>
  <summary>1 suggestion:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/patch-phase.md">patch-phase</a>

`patchPhase` should not be overridden, use `postPatch` instead.

Near pkgs/development/libraries/libcamera/default.nix:31:3:
```
   |
31 |   patchPhase = ''
   |   ^
```
</li>
  </ul>
</details>

**Comment by jansol** [comment]
Nice, I was hoping to see this packaged so pipewire can make use of it!

File: pkgs/development/libraries/libcamera/default.nix [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    python38Packages.jinja2
+
+    # IPA and signing
+    gnutls
+    openssl
+    boost
+
+    # gstreamer integration
+    gst_all_1.gstreamer
+    gst_all_1.gst-plugins-base
```
Comment (tomfitzhenry) [inline pkgs/development/libraries/libcamera/default.nix:48]: How does this integration work?

Is that idea that `libcamera` could be added to https://github.com/NixOS/nixpkgs/blob/3b6c3bee9174dfe56fd0e586449457467abe7116/pkgs/desktops/gnome/apps/cheese/default.nix#L81-L84 and Cheese would use libcamera as a gstreamer source?

Comment (CitadelCore) [reply pkgs/development/libraries/libcamera/default.nix:48]: I believe so, yeah. Check out https://libcamera.org/getting-started.html#using-gstreamer-plugin

File: pkgs/development/libraries/libcamera/default.nix [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    # cam integration
+    libevent
+
+    # qcam integration
+    qt5.qtbase
+    qt5.qttools
+    libtiff
+
+    # lttng tracing
+    lttng-ust
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/libcamera/default.nix:59]:
```suggestion
    # core requirements
    python3Packages.pyyaml
    python3Packages.ply
    python3Packages.jinja2
    # IPA and signing
    gnutls
    openssl
    boost
    # gstreamer integration
    gst_all_1.gstreamer
    gst_all_1.gst-plugins-base
    # cam integration
    libevent
    # qcam integration
    qt5.qtbase
    qt5.qttools
    libtiff
    # lttng tracing
    lttng-ust
```

File: pkgs/development/libraries/libcamera/default.nix [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+  ];
+
+  nativeBuildInputs = [
+    meson
+    ninja
+    pkg-config
+    python3
+    python38Packages.sphinx
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/libcamera/default.nix:67]:
```suggestion
    python3Packages.sphinx
```

File: pkgs/development/libraries/libcamera/default.nix [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+  meta = with lib; {
+    description = "An open source camera stack and framework for Linux, Android, and ChromeOS";
+    homepage = "https://libcamera.org";
+    license = licenses.bsd2;
+    maintainers = with maintainers; [ citadelcore ];
+  };
+
+  # Fixes error on a deprecated declaration
+  NIX_CFLAGS_COMPILE = [ "-Wno-error=deprecated-declarations" ];
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/libcamera/default.nix:84]:
```suggestion
  dontWrapQtApps = true;

  # Fixes error on a deprecated declaration
  NIX_CFLAGS_COMPILE = [ "-Wno-error=deprecated-declarations" ];

  meta = with lib; {
    description = "An open source camera stack and framework for Linux, Android, and ChromeOS";
    homepage = "https://libcamera.org";
    license = licenses.bsd2;
    maintainers = with maintainers; [ citadelcore ];
  };
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: pkgs/development/libraries/libcamera/default.nix (hunk 1 above)
Comment (tomfitzhenry) [reply pkgs/development/libraries/libcamera/default.nix:48]: Confirmed that gstreamer is able to find and use this.

File: pkgs/development/libraries/libcamera/default.nix [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+, python38Packages
+}:
+
+stdenv.mkDerivation {
+  pname = "libcamera";
+  version = "1.0.0";
+
+  src = fetchgit {
+    url = "git://linuxtv.org/libcamera.git";
+    rev = "040a6dcfbccaf1005da2c571769b1f89df1f07cf";
```
[the remark this answers is not in the document]
Comment (CitadelCore) [reply pkgs/development/libraries/libcamera/default.nix:27]: Unfortunately not, I checked and there aren't any tags.

File: pkgs/development/libraries/libcamera/default.nix [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+    sha256 = "sha256-RCNKFCv3fNabroCUR5MtKbyDS20J7leU8KkE1U5qe1M=";
+  };
+
+  patchPhase = ''
```
[the remark this answers is not in the document]
Comment (CitadelCore) [reply pkgs/development/libraries/libcamera/default.nix]: Changed.

File: pkgs/development/libraries/libcamera/default.nix (hunk 2 above)
Comment (CitadelCore) [reply pkgs/development/libraries/libcamera/default.nix]: Changed.

File: pkgs/development/libraries/libcamera/default.nix (hunk 3 above)
Comment (CitadelCore) [reply pkgs/development/libraries/libcamera/default.nix]: Changed.

File: pkgs/development/libraries/libcamera/default.nix (hunk 4 above)
Comment (CitadelCore) [reply pkgs/development/libraries/libcamera/default.nix]: Changed.

File: pkgs/development/libraries/libcamera/default.nix [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+, gst_all_1
+, graphviz
+, doxygen
+, python3
+, python3Packages
+}:
+
+stdenv.mkDerivation {
+  pname = "libcamera";
+  version = "1.0.0";
```
Comment (jtojnar) [inline pkgs/development/libraries/libcamera/default.nix:23]: In build files it listed as 0.0.0:

https://git.linuxtv.org/libcamera.git/tree/meson.build

So we should probably use the unstable version scheme.

File: pkgs/development/libraries/libcamera/default.nix [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+
+  dontWrapQtApps = true;
+
+  # Fixes error on a deprecated declaration
+  NIX_CFLAGS_COMPILE = [ "-Wno-error=deprecated-declarations" ];
+
+  meta = with lib; {
+    description = "An open source camera stack and framework for Linux, Android, and ChromeOS";
+    homepage = "https://libcamera.org";
+    license = licenses.bsd2;
```
Comment (jtojnar) [inline pkgs/development/libraries/libcamera/default.nix:82]: How did you determine this? I see `lgpl2Plus` in the build script: 

https://git.linuxtv.org/libcamera.git/tree/meson.build#n11

File: pkgs/development/libraries/libcamera/default.nix [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+    graphviz
+    doxygen
+  ];
+
+  mesonFlags = [ "-Dv4l2=true" ];
+
+  dontWrapQtApps = true;
+
+  # Fixes error on a deprecated declaration
+  NIX_CFLAGS_COMPILE = [ "-Wno-error=deprecated-declarations" ];
```
Comment (jtojnar) [inline pkgs/development/libraries/libcamera/default.nix:77]: This is an env var so it should be a string.

File: pkgs/development/libraries/libcamera/default.nix [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+  postPatch = ''
+    patchShebangs utils/
+  '';
+
+  buildInputs = [
+    # core requirements
+    python3Packages.pyyaml
+    python3Packages.ply
+    python3Packages.jinja2
+
```
Comment (jtojnar) [inline pkgs/development/libraries/libcamera/default.nix:40]: What about systemd for device hotplug enumeration?

File: pkgs/development/libraries/libcamera/default.nix [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+    pkg-config
+    python3
+    python3Packages.sphinx
+    graphviz
+    doxygen
+  ];
+
+  mesonFlags = [ "-Dv4l2=true" ];
+
+  dontWrapQtApps = true;
```
Comment (jtojnar) [inline pkgs/development/libraries/libcamera/default.nix:74]: I do not think anything will be wrapped when you do not add `wrapQtAppsHook` or the Qt `mkDerivation`.

And you would actually want the wrapping for the qcam app to work properly.

Though we might want to remove qt dependency from the default build to keep it lean, as qcam is mostly just a demo app IIUC.

File: pkgs/development/libraries/libcamera/default.nix [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+    sha256 = "sha256-RCNKFCv3fNabroCUR5MtKbyDS20J7leU8KkE1U5qe1M=";
+  };
+
+  postPatch = ''
+    patchShebangs utils/
+  '';
+
+  buildInputs = [
+    # core requirements
+    python3Packages.pyyaml
```
Comment (jtojnar) [inline pkgs/development/libraries/libcamera/default.nix:37]: These seem to be build-time requirements AFAICT so I would maybe try to move them inside `python3.withPackages` in `nativeBuildInputs`.

File: pkgs/development/libraries/libcamera/default.nix (hunk 11 above)
Comment (CitadelCore) [reply pkgs/development/libraries/libcamera/default.nix:74]: You raise a good point there about qcam just being a test app - I've decided to remove the qt dependencies and exclude it from the build.

File: pkgs/development/libraries/libcamera/default.nix (hunk 12 above)
Comment (CitadelCore) [reply pkgs/development/libraries/libcamera/default.nix:37]: For some reason `python3.withPackages` does not work at all here - I get `ModuleNotFoundError: No module named 'jinja2'`, but I can confirm if I at least move the packages into `nativeBuildInputs` the build still works. I've done this.

File: pkgs/development/libraries/libcamera/default.nix (hunk 9 above)
Comment (CitadelCore) [reply pkgs/development/libraries/libcamera/default.nix:77]: Changed.

File: pkgs/development/libraries/libcamera/default.nix (hunk 8 above)
Comment (CitadelCore) [reply pkgs/development/libraries/libcamera/default.nix:82]: Sorry, a mistake on my part. I've fixed this.

File: pkgs/development/libraries/libcamera/default.nix (hunk 7 above)
Comment (CitadelCore) [reply pkgs/development/libraries/libcamera/default.nix]: Done

File: pkgs/development/libraries/libcamera/default.nix (hunk 10 above)
Comment (CitadelCore) [reply pkgs/development/libraries/libcamera/default.nix]: Hmm, it's not listed as a dependency on libcamera itself - I'm not sure how to get it to compile with it either?

Comment (jtojnar) [reply pkgs/development/libraries/libcamera/default.nix]: It is listed in readme:

> for device hotplug enumeration: [optional]
> 	libudev-dev

(udev is part of systemd)

**Comment by jansol** [comment]
Any update on this?

Also please update to a more recent commit, there was a compatibility-breaking change somewhat recently that will prevent pipewire from building with an old libcamera starting from 0.3.33: https://gitlab.freedesktop.org/pipewire/pipewire/-/issues/1435

File: pkgs/development/libraries/libcamera/default.nix (hunk 10 above)
Comment (jansol) [reply pkgs/development/libraries/libcamera/default.nix]: https://libcamera.org/getting-started.html this page lists the optional dependencies

**Pull request merged**